enhancements/etcd/storage-migration-for-etcd-encryption.md

```diff
@@ -0,0 +1,182 @@
+  * Not required for bootstrapping or re-bootstrapping.
+  * No need for static pods.
+  * Will run on cluster.
+* Implement `openshift-kube-storage-version-migrator-operator`. 
+* Build `kube-storage-version-migrator` image.
+  * Fork `kubernetes-sigs/kube-storage-version-migrator` to `openshift`
+  * Enable CI on `openshift/kube-storage-version-migrator` repo.
+* Utilize `StorageVersionMigration` API when etcd storage encryption keys are updated.
+
+## User Stories [optional]
+
+* kube-apiserver-operator can request a resource migration and confirm when the migration has completed. 
+
+* openshift-apiserver-operator can request a resource migration and confirm when the migration has completed. 
+
+## Implementation Plan
+
+1. Implement `openshift-kube-storage-version-migrator-operator`.
+2. Have CVO manage `openshift-kube-storage-version-migrator-operator`.
+2. Add control loops to operators as needed.
+
+## Implementation Details/Notes/Constraints
+
+* Create a `openshift-kube-storage-version-migrator-operator`, managed by the CVO, to manage the resources needed to run the `kube-storage-version-migrator`.
+* Initiate a storage migration of resources, even if the storage version has not changed.
+* Monitor the progress of storage migration of resources.
+* Enable the `kube-apiserver-operator` to utilize the `StorageVersionMigration` API when managing etcd encryption keys.
+* Enable the `openshift-apiserver-operator` to utilize the `StorageVersionMigration` API when managing etcd encryption keys.
+* Inform the upstream `kube-storage-version-migrator` of our usage and our success/failure to gain influence for future decisions and create tests to ensure our use-cases.
+
+New control loop added to kube-apiserver-operator & openshift-apiserver-operator:
+
+```
+for { // control loop
+
+    key := getNextWriteKey()
+
+    for resource := range encryptedResources {
+        migration := getStorageVersionMigrationFor(resource, key)
+        if migration not found {
+            createStorageVersionMigrationFor(resource, key)
+        }
+    }
+
+    if migrationsAreCompleteFor(key) {
+        setOperatorStatusPending(false)
+        enableWriteKey(key)
+    } else {
+        setOperatorStatusPending(false)
+    }
+        
+}
+```
+
+## Risks and Mitigations
+
+* kube-storage-version-migrator is currently in alpha with a clear plan to beta.
+
+* If encryption keys change while api-servers are a different levels, forcing a storage migration might render some api-servers unable to read the etcd data. To mitigate, we don’t request storage migrations unless all control planes are at the same level.
+
+* In order to insure that storage migrations initiated by other components. For etcd encryption, for example, we distribute and maintain multiple read keys allowing for a storage migration to “occur” even when not all control planes have the latest shared write key.
+
```

Comment:
actually I would drop both risks. They are not risks of this proposal, but inherent to the consumers of the migration tool.

Reply:
Done